gr-soapy/lib/block_impl.h

```diff
@@ -0,0 +1,274 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2021 Jeff Long
+ * Copyright 2018-2021 Libre Space Foundation <http://libre.space>
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#ifndef INCLUDED_GR_SOAPY_BLOCK_IMPL_H
+#define INCLUDED_GR_SOAPY_BLOCK_IMPL_H
+
+#include <gnuradio/soapy/block.h>
+
+#include <SoapySDR/Device.hpp>
+#include <SoapySDR/Modules.hpp>
+#include <SoapySDR/Registry.hpp>
+#include <SoapySDR/Version.hpp>
+
+namespace gr {
+namespace soapy {
+
+using cmd_handler_t = std::function<void(pmt::pmt_t, size_t)>;
+
+/*!
+ * \brief Base block implementation for SDR devices.
+ */
+
+class block_impl : virtual public block
+{
+private:
+    const int d_direction;
+    const std::string d_dev_str;
+    const std::string d_args;
+
+    size_t d_mtu = 0;
+    size_t d_nchan;
+    std::string d_stream_args;
+    std::vector<size_t> d_channels;
+    std::string d_soapy_type;
+    std::map<pmt::pmt_t, cmd_handler_t> d_cmd_handlers;
+    std::vector<SoapySDR::Kwargs> d_tune_args;
+
+    void register_msg_cmd_handler(const pmt::pmt_t& cmd, cmd_handler_t handler);
+
+    /*!
+     * \brief Raise std::invalid_argument if channel is invalid
+     */
+    void validate_channel(size_t channel);
+
+protected:
+    block_impl(int direction,
+               const std::string& device,
+               const std::string& type,
+               size_t nchan,
+               const std::string& dev_args,
+               const std::string& stream_args,
+               const std::vector<std::string>& tune_args,
+               const std::vector<std::string>& other_settings);
+    block_impl(const block_impl&) = delete;
+    block_impl& operator=(const block_impl&) = delete;
+    ~block_impl();
```

Comment:
I'm nut sure, but does the destructor here need to be marked virtual?

Reply:
assert on `d_stream`: `start()` and `stop()` are called from the block executor cons/destructors, so a double-start is a major framework problem.